src/transformers.jl

```diff
@@ -0,0 +1,66 @@
+
+"""
+    Transformation
+
+Abstract supertype for all transformations.
+"""
+abstract type Transformation end
+
+# Make Transforms callable types
+(t::Transformation)(x; kwargs...) = transform(x, t; kwargs...)
+
+"""
+    transform!(data::T, transformation::Transformation; kwargs...) -> T
+
+Apply `transformation` to the `data` in-place.
+Where possible, this should be extended for new data types `T`.
+"""
+function transform! end
+
+"""
+    transform(data::T, transformation::Transformation; kwargs...) -> T
+
+Non-in-place version of [`transform!`](@ref), which it delegates to by default.
+Does not need to be extended unless in-place transformation is not possible.
+"""
+function transform end
+
+"""
+    transform!(A::AbstractArray{T}, ::Transformation; dims=:, kwargs...) where T <: Real
+
+Applies the transformation to each element of `A`.
+Optionally specify the `dims` to apply the transformation along certain dimensions.
+"""
+function transform!(
+    A::AbstractArray{T}, t::Transformation; dims=:, kwargs...
+) where T <: Real
+    dims == Colon() && return _transform!(A, t; kwargs...)
+
+    return map(eachslice(A; dims=dims)) do x
+        return _transform!(x, t; kwargs...)
+    end
+end
+
+transform(x, t::Transformation; kwargs...) = transform!(_trycopy(x), t; kwargs...)
+
+"""
+    transform!(table::T, ::Transformation; cols=nothing)::T where T
+
+Applies the transformation to each of the specified columns in the `table`.
+If no `cols` are specified, then the transformation is applied to all columns.
+"""
+function transform!(table::T, t::Transformation; cols=nothing)::T where T
```

Comment:
I haven't seen `f()::T` syntax before - is that specifying the return type?
Why doesn't the Array transform have that?

Reply:
It does indeed specify the return type. It's not required but will raise an error if the function returns something not of type T

Reply:
> Why doesn't the Array transform have that?

So is it because there is particular risk of mutating the type with tables compared to arrays?

Reply:
Yeah, for instance certain operations might return a `SubDataFrame`, `DataFrameRow` or just extract the raw data and return a `Matrix`. We'd want to prevent this from happening.